concepts/pivoted/eICU-pivoted-OASIS.sql

```diff
@@ -0,0 +1,479 @@
+)
+
+, cohort_OASIS AS (
+  SELECT cohort.patientunitstayid, 
+  pre_ICU_LOS_OASIS.pre_ICU_LOS_OASIS, 
+  age_OASIS.age_OASIS, 
+  GCS_OASIS.GCS_OASIS,
+  heartrate_OASIS.heartrate_OASIS,
+  MAP_OASIS.MAP_OASIS,
+  respiratoryrate_OASIS.respiratoryrate_OASIS,
+  temperature_OASIS.temperature_OASIS,
+  urineoutput_OASIS.urineoutput_OASIS,
+  VENT_OASIS.VENT_OASIS,
+  electivesurgery_OASIS.electivesurgery_OASIS
+  FROM  `physionet-data.eicu_crd.patient` AS cohort
+
+  LEFT JOIN pre_ICU_LOS_OASIS
+  ON cohort.patientunitstayid = pre_ICU_LOS_OASIS.pid_LOS
+
+  LEFT JOIN age_OASIS
+  ON cohort.patientunitstayid = age_OASIS.pid_age 
+
+  LEFT JOIN GCS_OASIS
+  ON cohort.patientunitstayid = GCS_OASIS.pid_GCS 
+
+  LEFT JOIN heartrate_OASIS
+  ON cohort.patientunitstayid = heartrate_OASIS.pid_HR 
+
+  LEFT JOIN MAP_OASIS
+  ON cohort.patientunitstayid = MAP_OASIS.pid_MAP 
+
+  LEFT JOIN respiratoryrate_OASIS
+  ON cohort.patientunitstayid = respiratoryrate_OASIS.pid_RR
+
+  LEFT JOIN temperature_OASIS
+  ON cohort.patientunitstayid = temperature_OASIS.pid_temp
+
+  LEFT JOIN urineoutput_OASIS
+  ON cohort.patientunitstayid = urineoutput_OASIS.pid_urine
+
+  LEFT JOIN VENT_OASIS
+  ON cohort.patientunitstayid = VENT_OASIS.pid_vent
+
+  LEFT JOIN electivesurgery_OASIS
+  ON cohort.patientunitstayid = electivesurgery_OASIS.pid_adm
+
+)
+
+, score_impute AS (
+
+SELECT cohort_OASIS.*,
+
+  IFNULL(pre_ICU_LOS_OASIS, 0) AS pre_ICU_LOS_OASIS_Imp,
+  IFNULL(age_OASIS, 0) AS age_OASIS_Imp, 
+  IFNULL(GCS_OASIS, 0) AS GCS_OASIS_Imp, 
+  IFNULL(heartrate_OASIS, 0) AS heartrate_OASIS_Imp,
+  IFNULL(MAP_OASIS, 0) AS MAP_OASIS_Imp,
+  IFNULL(respiratoryrate_OASIS, 0) AS respiratoryrate_OASIS_Imp,
+  IFNULL(temperature_OASIS, 0) AS temperature_OASIS_Imp, 
+  IFNULL(urineoutput_OASIS, 0) AS urineoutput_OASIS_Imp, 
+  IFNULL(VENT_OASIS, 0) AS VENT_OASIS_Imp, 
+  IFNULL(electivesurgery_OASIS, 0) AS electivesurgery_OASIS_Imp, 
+
+FROM cohort_OASIS
+)
+
+--Compute overall score
+-- OASIS_NULL -> only cases where all components have a Non-NULL value
+-- OASIS_Imp -> Imputation in case of NULL values, with 0's (common approach for severity of illness scores)
+, score AS (
+SELECT patientunitstayid, 
+    MAX(pre_ICU_LOS_OASIS) AS pre_ICU_LOS_OASIS,
+    MAX(age_OASIS) AS age_OASIS,
+    MAX(GCS_OASIS) AS GCS_OASIS,
+    MAX(heartrate_OASIS) AS heartrate_OASIS,
+    MAX(MAP_OASIS) AS MAP_OASIS,
+    MAX(respiratoryrate_OASIS) AS respiratoryrate_OASIS,
+    MAX(temperature_OASIS) AS temperature_OASIS,
+    MAX(urineoutput_OASIS) AS urineoutput_OASIS,
+    MAX(VENT_OASIS) AS VENT_OASIS,
+    MAX(electivesurgery_OASIS) AS electivesurgery_OASIS,
+    MAX(pre_ICU_LOS_OASIS + 
+        age_OASIS + 
+        GCS_OASIS + 
+        heartrate_OASIS + 
+        MAP_OASIS + 
+        respiratoryrate_OASIS + 
+        temperature_OASIS + 
+        urineoutput_OASIS + 
+        VENT_OASIS + 
+        electivesurgery_OASIS) AS OASIS_NULL,
```

Comment:
Are you sure this works? What happens when one of the values is `NULL`? My suspicion is this ends up being identical to `OASIS_Imp`, or worse it becomes `NULL` if any component is `NULL`.

Reply:
Yes, it will return a 'NULL' for the score if a single column contains a 'NULL'. That is why we went for the imputation approach, as a convenience for both us and future users.

Reply:
So I'll leave this up to you, but ultimately, I think if I'm using this table I want one of two things:

- the original scores for each component, with `NULL` if no score could be created
- the final tabulated score (OASIS) with `NULL` imputed as a score of 0

I don't see anyone realistically using `OASIS_NULL`. If you think it's useful to provide a column indicating you imputed data, then maybe it makes sense to just directly do that? That is, you end up with:

- the component score columns
- a column containing a 0 if all score values were tabulated, or a 1 if any score was missing
- the OASIS score using imputation for `NULL`

Reply:
creating a separate column would have been quite complicated
instead I just kept the score's components with 'NULL' and the final score OASIS where I replaced them with '0'